### DISCUSSION TITLE
USB devices occasionally reconnecting == normal macOS behavior???

### DISCUSSION THREAD

**Discussion by tigersoul925** [Hammerspoon/hammerspoon#3443] [Q&A]
I stumbled upon something I wouldn't even have noticed if it wasn't for the fact I was checking another USB thing, but it seems some USB devices in my system occasionally disconnect and then reconnects within 0-1 secs. This is happening while leaving the connected device as it seems entirely unaffected. I've got an external LaCie disk and a USB stick that I've noticed it happening for and maybe a network USB interface.

When it happens, the disk just goes on as if nothing happened. It's not unmounted or ejected. I've even been lucky enough to have a binary read check running while this occurred and it was 100% unaffected. I just can't see how a USB disk could ever be actually disconnected and reconnected without operation being affected. So what is this? normal macOS behavior???

Has anyone else noticed this? It can be monitored with code something like this:

 ```
 usbConnectsTimer = hs.timer.new(3, function()
    usbConnectsTimer:stop()
    print("> usb connect for devices: " .. usbConnectsDevs)
    usbConnects = 0
    usbConnectsDevs = nil
  end)

  usbDisconnectsTimer = hs.timer.new(3, function()
    usbDisconnectsTimer:stop()
    print("> usb disconnect for devices: " .. usbDisconnectsDevs)
    usbDisconnects = 0
    usbDisconnectsDevs = nil
  end)

  usbConnects = 0
  usbDisconnects = 0
  usbDCWatcher = {}

  usbMonitor = hs.usb.watcher.new(function(info)
    -- iNotify(log,"USB Monitor: " .. info['eventType'] .. " " .. info['vendorName'] .. " " .. info['productName'])
    local event = info['eventType']
    local vendor = info['vendorName']
    local product = info['productName']
    print("USB Monitor: " .. event .. " " .. vendor .. " " .. product)
    if event == "added" then
      if usbDebugAllConns then
        usbConnectsTimer:setNextTrigger(3)
        usbConnects = usbConnects + 1
        if usbConnectsDevs ~= nil then usbConnectsDevs = usbConnectsDevs .. ", " .. vendor .. " " .. product else usbConnectsDevs = vendor .. " " .. product end
      end
      if usbDebugReconns then
        currTime = hs.timer.secondsSinceEpoch()
        for i, v in pairs(usbDCWatcher) do
          if currTime-v>30 then usbDCWatcher[i] = nil end
        end
        if usbDCWatcher[vendor .. " " .. product] ~= nil then
          print("USB Device reconnect occured: " .. vendor .. " " .. product)
        end
      end
    elseif event == "removed" then
      if usbDebugAllConns then
        usbDisconnectsTimer:setNextTrigger(3)
        usbDisconnects = usbDisconnects + 1
        if usbDisconnectsDevs ~= nil then usbDisconnectsDevs = usbDisconnectsDevs .. ", " .. vendor .. " " .. product else usbDisconnectsDevs = vendor .. " " .. product end
      end
      if usbDebugReconns then usbDCWatcher[vendor .. " " ..product]=hs.timer.secondsSinceEpoch() end
    else print("Unknown USB event in usb watcher") end
  end):start()

  function getUsbDevices()
    local devs = hs.usb.attachedDevices()
    local cnt = 0
    local res = {}
    local vendor
    for i,v in pairs(devs) do
      cnt = cnt + 1
      if v['vendorName'] ~= nil then vendor = v['vendorName'] else vendor = "(no vendor)" end
      if v['productName'] ~= nil then product = v['productName'] else product = "(no productname)" end
      table.insert(res,vendor .. " " .. product)
    end
    table.sort(res)
    print("\n" .. table.concat(res,"\n"))
    print(cnt .. " devices attached")
  end

```

**Comment by andrewcrook** [comment]
could you maybe write a bash script to loop and print/log the diff the output of something like

`ioreg -p IOUSB -w0 | sed 's/[^o]*o //; s/@.*$//' | grep -v '^Root.*'`
```
AppleT8103USBXHCI
IOUSBHostDevice
USB Receiver
USB Controls
AppleT8103USBXHCI
```
it may prove that the issue is with hammerspoon or code?
I have had strange things happen with watchers for example being triggered twice in a row rather than the once. I found that I either had to define and set a “isrunning” variable and test it to stop code from being running whilst the same callback is already running or stop and restart the watcher whilst the callback is running.

**Reply by tigersoul925** [reply]
I have already verified that this is not Hammerspoon specific. BTT also logs USB connect/disconnect and it mirrors the events. The script is a good idea, but the reconnect often happens with the connect and disconnect event happening the same second. It's a very brief event.

**Reply by andrewcrook** [reply]
I must admit I haven't had such issues with connectivity in general since moving to Apple Silicon (from Intel Macs). Saying that, I don't normally have USB storage connected. So I will have to test it. I do know M1 Mac's Thunderbolt 4 doesn't fully support USB 3.1 gen 2 causing devices to run slower and some do act oddly.  I will try to replicate your issue later.

**Reply by tigersoul925** [reply]
> I must admit I haven't had such issues with connectivity in general since moving to Apple Silicon. Saying that, I don't normally have USB storage connected. So I will have to test it. I do know M1 Mac's Thunderbolt 4 doesn't fully support USB 3.1 gen 2 causing devices to run slower and some do act oddly. I will try to replicate your issue later.

Yeah, I'm still stumped at what they are coming from. It seems one of my storages, a LaCie disk, is seeing the most events. But it's also the most used. The events are relatively rare and hard to connect to something specific. Guess I should just be happy they don't seem to affect the data stream from the drive, but it's still weird and I don't feel comfy exactly.
